src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,292 @@
+      # To get always the same solution one could require that v*Q*v1 > 0, but this has very huge impact on the efficiency. 
+      @vprintln :Vinberg 2 "direction vector v1 = $v1"
+      return v1
+    end
+    signal += 1
+  end
+  @req signal < 10000000 "Choose another v0" # break the algorithm if no v1 was found
+end
+
+@doc raw"""
+    _check_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix}, direction_vector::ZZMatrix) -> ZZMatrix
+
+We check if the given direction vector holds the following requirements:
+- v0.v1 = 0
+- v.v1 != 0 for all possible roots v with v.v0 = 0
+"""
+function _check_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix}, direction_vector::ZZMatrix)
+  v1 = transpose(direction_vector)
+  @req (v0*Q*v1)[1, 1] == 0 "The direction vector has to be orthogonal on v0"
+  @req all((v*Q*v1)[1, 1] != 0 for v in possible_vec) "The direction vector cannot be orthogonal on one of the possible roots"
+  return v1
+end
+
+@doc raw"""
+    vinberg_algorithm(Q::ZZMatrix, upper_bound::ZZRingElem; v0::ZZMatrix, root_lengths::Vector{ZZRingElem}, direction_vector::ZZMatrix) -> Vector{ZZMatrix}
+
+Return the roots r of a given hyperbolic reflection lattice Q with squared length contained in root_lengths and 
+by increasing order of the value ((r.v0)^2)//(r^2), stopping by upper_bound.
+If root_lengths is not defined it takes all possible values of r^2.
+If v0 lies on several possible roots and if there is no given direction vector it is a random choice which reflection chamber
+next to v0 will be computed.
+
+# Arguments
+- 'Q': symmetric ZZ matrix 
+- 'upper_bound': the integer upper bound of the value ((r.v0)^2)//(r^2)
+- 'v0': row vector with $v^2 > 0$
+- 'root_lengths': the possible integer values of r^2
+- 'direction_vector': row vector v1 with v0.v1 = 0 and v.v1 != 0 for all possible roots v with v.v0 = 0
+
+The output is given in the ambient representation.
+"""
+function vinberg_algorithm(Q::ZZMatrix, upper_bound::ZZRingElem; v0 = matrix(ZZ, 1, 1, [0])::ZZMatrix, root_lengths=ZZRingElem[]::Vector{ZZRingElem}, direction_vector = matrix(ZZ, 1, 1, [0])::ZZMatrix)
+  @req is_symmetric(Q) "Matrix is not symmetric"
+  v0 = _check_v0(Q, v0)
+  if isempty(root_lengths)
+    real_root_lengths = _all_root_lengths(Q)
+  else
+    real_root_lengths = _check_root_lengths(Q, root_lengths) # sort out impossible lengths
+  end
+  iteration = _distance_indices(upper_bound, real_root_lengths) # find the right order to iterate through v.v0 and v^2
+  roots = _distance_0(Q, v0, real_root_lengths, direction_vector) # special case v.v0 = 0
+  for (n, k) in iteration # we  search for vectors which solve n = v.v0 and k = v^2
+    possible_Vec = Oscar.short_vectors_affine(Q, v0, QQ(n), k)
+    for v in possible_Vec
+      if !_is_primitive(v)
+        continue
+      end
+      if _crystallographic_condition(Q, v)
+        if _check_coorientation(Q, roots, v, v0)
```

Comment:
This is the hot loop. So these two function will be called many times. Can you make them non-allocating?